overlord/devicestate/remodel_test.go

```diff
@@ -931,3 +945,377 @@ func (s *remodelLogicSuite) TestReregRemodelContextNewSerial(c *C) {
+		// this is running as part of post finish step, so the state has
+		// already been updated
+		serial, err = s.mgr.Serial()
+		c.Assert(err, IsNil)
+		c.Check(serial.Model(), Equals, "other-model")
+		c.Check(serial.Serial(), Equals, "serialserialserial2")
+		return nil
+	})
+	s.AddCleanup(restore)
+
+	// finish fails because we haven't set the seed system label yet
+	err = remodCtx.Finish()
+	c.Assert(err, ErrorMatches, "internal error: recovery system label is unset during remodel finish")
+	c.Check(resealKeysCalls, Equals, 0)
+
+	// set the label internally
+	devicestate.RemodelSetRecoverySystemLabel(remodCtx, "1234")
+	err = remodCtx.Finish()
+	c.Assert(err, IsNil)
+	c.Check(resealKeysCalls, Equals, 2)
+
+	var seededSystemsFromState []map[string]interface{}
+	err = s.state.Get("seeded-systems", &seededSystemsFromState)
+	c.Assert(err, IsNil)
+	c.Assert(seededSystemsFromState, HasLen, 2)
+	c.Assert(seededSystemsFromState[1], DeepEquals, map[string]interface{}{
+		"system":    "0000",
+		"model":     "my-model",
+		"brand-id":  "my-brand",
+		"revision":  float64(0),
+		"timestamp": s.oldModel.Timestamp().Format(time.RFC3339Nano),
+		"seed-time": s.oldSeededTs.Format(time.RFC3339Nano),
+	})
+	// new system is prepended, since timestamps are involved clear ones that weren't mocked
+	c.Assert(seededSystemsFromState[0]["seed-time"], FitsTypeOf, "")
+	newSeedTs, err := time.Parse(time.RFC3339Nano, seededSystemsFromState[0]["seed-time"].(string))
+	c.Assert(err, IsNil)
+	seededSystemsFromState[0]["seed-time"] = ""
+	c.Assert(seededSystemsFromState[0], DeepEquals, map[string]interface{}{
+		"system":    "1234",
+		"model":     "other-model",
+		"brand-id":  "my-brand",
+		"revision":  float64(0),
+		"timestamp": newModel.Timestamp().Format(time.RFC3339Nano),
+		"seed-time": "",
+	})
+	c.Assert(newSeedTs.After(s.oldSeededTs), Equals, true)
+}
+
+func (s *uc20RemodelLogicSuite) TestUpdateRemodelContext(c *C) {
+	modelDefaults := make(map[string]interface{}, len(uc20ModelDefaults))
+	for k, v := range uc20ModelDefaults {
+		modelDefaults[k] = v
+	}
+	// simple model update with bumped revision
+	modelDefaults["revision"] = "2"
+	newModel := s.brands.Model("my-brand", "my-model", modelDefaults)
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	m, err := boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	// the system has already been promoted
+	m.CurrentRecoverySystems = append(m.CurrentRecoverySystems, "1234")
+	m.GoodRecoverySystems = append(m.GoodRecoverySystems, "1234")
+	c.Assert(m.Write(), IsNil)
+
+	remodCtx, err := devicestate.RemodelCtx(s.state, s.oldModel, newModel)
+	c.Assert(err, IsNil)
+	c.Check(remodCtx.ForRemodeling(), Equals, true)
+	c.Check(remodCtx.Kind(), Equals, devicestate.UpdateRemodel)
+	groundCtx := remodCtx.GroundContext()
+	c.Check(groundCtx.ForRemodeling(), Equals, false)
+	c.Check(groundCtx.Model().Revision(), Equals, 0)
+	c.Check(groundCtx.Store, PanicMatches, `retrieved ground context is not intended to drive store operations`)
```

Comment:
Hmmm, not groundCtx.Store**()** ?

Reply:
No, this one takes whatever matches `func()`